walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,73 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Set<String> allAttendees = new HashSet<>(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+    
+    Collection<TimeRange> ranges = getTimeRanges(events, allAttendees, request.getDuration());
+    if (ranges.size() == 0 && mandatoryAttendees.size() > 0) {
+      ranges = getTimeRanges(events, mandatoryAttendees, request.getDuration());
+    }
+    return ranges;
+  }
+
+  public Collection<TimeRange> getTimeRanges(Collection<Event> events, Collection<String> attendees, long duration) {
+    Comparator<Event> comparator = (Event e1, Event e2) -> {
+        return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+    };     
+    /**
+     * All attendees are treated as mandatory.
+     * Events are sorted in chronological order with respect to start times.
+     * When a new event is encountered, check if there is an overlap with the meeting
+     * attendees. If so, add a meeting timespan from the current start to the start of this
+     * event if it is a valid length of time. Then set the current start to end of this event.
+     */
+    boolean rejectedEvent = false;
+    List<Event> l = new ArrayList(events);
```

Comment:
nit: better naming for "l", so it's easier to remember what it is later in the code. "sortedEvents" perhaps?